CT-1243 ability to add deactivated users to a team

aliuk2012 opened the pull request:
ATM, when you try to add a deactivated user to a team. It reports
that an existing user has been added to the team but it does not list
them in the teams users list. The reason for this is because the user
is still soft deleted.

This commit removes the soft delete from the users account
if the user is deactivated

misaka commented on app/controllers/cases_controller.rb:
```diff
@@ -29,7 +29,8 @@ def create
 
-  def show;
+  def show
+    set_permitted_events
```
We should test that this bit works.

misaka replied:
This is being tested now.

misaka commented on app/controllers/cases_controller.rb:
```diff
@@ -50,6 +51,7 @@ def close
     authorize @case, :can_close_case?
 
     @case.close(current_user.id)
+    set_permitted_events
```
We should test that this bit works.

misaka replied:
This is being tested now, I see.

EdwardAndress commented:
OK, this is done! @misaka and @aliuk2012

EdwardAndress commented:
@misaka I suspect you will want to re-review this

aliuk2012 commented on lib/case_state_machine.rb:
```diff
@@ -20,10 +20,28 @@ class CaseStateMachine
   end
 
+  event :upload_response do
+    transition from: :drafting, to: :drafting
```
Why are we transitioning from draft to drafting?

EdwardAndress replied:
Because we're using `#available_events` in cases_controller.rb to set @permitted_events, which is then used to determine which buttons are shown.  We need, therefore, an event for every button (`#upload_response` in this example) but there's no state change associated with this part of the process, as yet, so there's no state change triggered by this event.  This transition also generates a `CaseTransition`, which allows us to record the `upload_response` event in the case history.

Also... this does control when we can trigger the `#upload_response` event, because of the `from:`

There may be a nicer way of writing this bit of code though...

aliuk2012 replied:
Thanks for the explanation. That makes sense for the time being keeping the state as "Drafting" by transitioning to drafting and recording something has happened makes sense. I guess this is likely to change once we start adding trigger cases and the approval progress or maybe when we want to show DACU that their has been some form of work being done on the case.

misaka commented on db/structure.sql:
```diff
@@ -547,13 +547,6 @@ CREATE UNIQUE INDEX index_cases_on_number ON cases USING btree (number);
--- Name: index_cases_on_requester_type; Type: INDEX; Schema: public; Owner: -
---
-
-CREATE INDEX index_cases_on_requester_type ON cases USING btree (requester_type);
```
Is this right? Do we want to be removing this index?

EdwardAndress replied:
@misaka - no, we don't.  There was something not quite right with my structure.sql file and I had to drop/create/migrate to sort it out.  Thanks!

misaka commented on app/models/case.rb:
```diff
@@ -129,6 +129,12 @@ def responder_assignment_accepted(assignee_id)
+  def add_responses(assignee_id, responses)
+    self.attachments << responses
+    filenames = responses.map(&:filename)
+    state_machine.upload_response!(assignee_id, filenames)
```
So `add_response` triggers an `upload_response` event. Does that sound right? Sounds like this could be confusing and we should probably use the same name here.

EdwardAndress replied:
OK, @misaka, not sure if what I have done is exactly what you meant but, I have left the controller as `#upload_responses` because there are file actually being uploaded.  Downstream from there, I have consistently used `#add_responses`, because from there onwards, the files have been uploaded and we are just manipulating records to associate `CaseAttachment`s with `Case`s etc

misaka replied:
Close enough. I wasn't sure because I imagined one day we'd add a way to attach canned responses to cases, so I thought that differentiating from uploaded reponses might be good, but I think this change makes sense and we'll cross that bridge when we get there.

misaka commented on app/helpers/cases_helper.rb:
```diff
@@ -1,15 +1,27 @@
+          class: 'button', method: :patch
+    when :accept_responder_assignment, :reject_responder_assignment
+      link_to I18n.t('common.case.accept_or_reject'),
+          edit_case_assignment_path(@case, @case.assignments.last),
```
This worries me. It worried me before, and having looked at it a couple times it still worries me. I think I see why now, because one day we're going to have more than one kind of assignment here (responders and caseworkers, right?), and at that point we risk having a subtle/intermittent bug show up here that might prove hard to track down. I think that at the very least, we should only be looking at responder assignments, but depending on how tricky it is (and paranoid we are), we may want to filter it to only those assignments that match the current user's id (could we ever have multiple drafters assigned with a first-to-respond-gets-it strategy?).

EdwardAndress replied:
I've fixed this by implementing a new method on my current branch, for CT-421.

misaka replied:
👍 Sounds good.

misaka commented on spec/controllers/cases_controller_spec.rb:
```diff
@@ -61,6 +61,45 @@
+        let(:accepted_case)         { create(:accepted_case)   }
+        before { get :show, params: { id: accepted_case.id   } }
+
+        it 'permitted_events == []' do
```
This doesn't match the test below, assigners can (for now) add responses.

EdwardAndress replied:
thanks!

EdwardAndress replied:
done

misaka commented:
👍 Looking good to me!

EdwardAndress commented:
Cool, thanks, it's going in now.

aliuk2012 committed:
CT-1243 ability to add deactivated users to a team

ATM, when you try to add a deactivated user to a team. It reports
that an existing user has been added to the team but it does not list
them in the teams users list. The reason for this is because the user
is still soft deleted.

This commit removes the soft delete from the users account
if the user is deactivated
